pwndbg/heap/ptmalloc.py

```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+                        # Try to find `ldr.w reg, [pc, #offset]`, then `add reg, pc`
+                        if not ldrw_instr:
+                            if instr.mnemonic == 'ldr.w':
+                                ldrw_instr = instr
+                        else:
+                            reg = ldrw_instr.operands[0].str
+                            if instr.mnemonic == 'add' and instr.op_str == reg + ", pc":
+                                # ldr.w reg, [pc, #offset]
+                                offset = ldrw_instr.operands[1].mem.disp
+                                offset = pwndbg.memory.s32((ldrw_instr.address + 4 & -4) + offset)
+                                # add reg, pc
+                                self._main_arena_addr = offset + instr.address + 4
+                else:
+                    inform_report_issue("main_arena")
+                    raise OSError("Cannot find the symbol via heuristics")
+            # try to search main_arena in .data of libc if we can't find it via above trick
+            if not self._main_arena_addr:
+                _IO_2_1_stdin_addr = pwndbg.symbol.address('_IO_2_1_stdin_')
+                _IO_list_all_addr = pwndbg.symbol.address('_IO_list_all')
+                # main_arena is between _IO_2_1_stdin and _IO_list_all
+                for addr in range(_IO_2_1_stdin_addr, _IO_list_all_addr, pwndbg.arch.ptrsize):
+                    tmp_arena = self.malloc_state(addr)
+                    if tmp_arena["next"] == addr:
+                        self._main_arena_addr = addr
+                        break
+                if not self._main_arena_addr:
+                    # there are more than one arena, try to find by main_arena.top and main_arena.max_system_mem
+                    heap_page = next(x for x in pwndbg.vmmap.get() if "heap]" in x.objfile)
+                    for addr in range(_IO_2_1_stdin_addr, _IO_list_all_addr, pwndbg.arch.ptrsize):
+                        tmp_arena = self.malloc_state(addr)
+                        if heap_page.start <= tmp_arena["top"] <= heap_page.end:
+                            if tmp_arena["max_system_mem"] != 0:
+                                self._main_arena_addr = addr
+                                break
+
+        if self._main_arena_addr:
+            self._main_arena = self.malloc_state(self._main_arena_addr)
+        else:
+            inform_report_issue("main_arena")
+            raise OSError("Cannot find the symbol via heuristics")
+
+        return self._main_arena
+
+    def has_tcache(self):
+        # TODO/FIXME: Can we determine the tcache_bins existence more reliable?
+
+        # There is no debug symbols, we determine the tcache_bins existence by checking glibc version only
+        return self.is_initialized() and pwndbg.glibc.get_version() >= (2, 26)
+
+    def get_tls_base_via_errno_location(self):
+        assert pwndbg.arch.current in ("x86-64", "i386", "arm")
```

Comment:
@lebr0nli can we add here some comment about Aarch64, as in why we don't have it here?

From https://maskray.me/blog/2021-02-14-all-about-thread-local-storage it seems TLS exists on aarch64... (well, why would it not :D)

Reply:
Yeah, aarch64 has TLS, too.
But the `TPIDR_EL0` register which is used for aarch64 TLS seems always work.
So I didn’t implement that for it.
Is it possible that `TPIDR_EL0` failed to lead us to find TLS base?
If the answer is yes, maybe we need to implement for aarch64, too.

For other architectures:
I use this function for arm32 because I can’t find a way to access the coprocessor register(if I didn't misunderstand it).
And I use this for x86/x64 because sometimes we can’t always find gsbase/fsbase since its implementation is limited.